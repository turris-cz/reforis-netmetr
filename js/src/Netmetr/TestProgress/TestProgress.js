/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";

import useNetmetrTest from "./hooks";
import "./TestProgress.css";

TestProgress.propTypes = {
    ws: PropTypes.object.isRequired,
    asyncId: PropTypes.string,
};

export default function TestProgress({ ws, asyncId }) {
    const [data] = useNetmetrTest(ws, asyncId);

    if (!data) return null;

    return (
        <>
            <h4 className="text-center">
                {_("Test progress")}
            </h4>
            <div className="progress">
                {/* eslint-disable-next-line jsx-a11y/control-has-associated-label */}
                <div
                    id="progressbar"
                    className="progress-bar progress-bar-striped progress-bar-animated bg-success"
                    role="progressbar"
                    aria-valuenow={data.percent}
                    aria-valuemin="0"
                    aria-valuemax="100"
                    style={{ width: `${data.percent}%` }}
                >
                    {data.msg}
                </div>
            </div>
        </>
    );
}
