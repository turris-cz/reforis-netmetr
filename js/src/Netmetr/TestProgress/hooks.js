/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import { useState, useEffect } from "react";
import { useAlert, useWSForisModule } from "foris";

export default function useNetmetrTest(ws) {
    const [data, setData] = useState(null);
    const [setAlert] = useAlert();

    const [dataProcess] = useWSForisModule(ws, "netmetr", "measure_and_download_data_notification");
    useEffect(() => {
        if (dataProcess) {
            setData((currentData) => ({ ...currentData, ...dataProcess }));
        }
    }, [dataProcess]);

    const [dataFinished] = useWSForisModule(ws, "netmetr", "measure_and_download_data_finished");
    useEffect(() => {
        if (dataFinished) {
            setData(() => null);
        }
    }, [dataFinished, setAlert]);

    return [data];
}
