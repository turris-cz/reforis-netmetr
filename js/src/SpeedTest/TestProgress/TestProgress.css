/*
 * Copyright (C) 2020-2021 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

.progress {
    height: 35px;
}

.progress-bar-striped {
    background-size: 1rem 1rem;
}

@keyframes progress-bar-stripes {
    from {
        background-position: 1rem 0;
    }
    to {
        background-position: 0 0;
    }
}
