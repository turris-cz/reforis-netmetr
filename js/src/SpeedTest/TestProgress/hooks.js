/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import { useState, useEffect } from "react";
import { useAlert, useWSForisModule } from "foris";

export default function useNetmetrTest(ws, asyncId) {
    const [data, setData] = useState(null);
    const [setAlert] = useAlert();

    const [measureAndDownloadData] = useWSForisModule(ws, "netmetr", "measure_and_download_data_notification");
    useEffect(() => {
        if (measureAndDownloadData && measureAndDownloadData.async_id === asyncId) {
            setData((currentData) => ({ ...currentData, ...measureAndDownloadData }));
        }
    }, [asyncId, measureAndDownloadData]);

    const [measureAndDownloadDataFinished] = useWSForisModule(ws, "netmetr", "measure_and_download_data_finished");
    useEffect(() => {
        if (measureAndDownloadDataFinished && measureAndDownloadDataFinished.async_id === asyncId) {
            setData(() => null);
        }
    }, [asyncId, measureAndDownloadDataFinished, setAlert]);

    return [data];
}
