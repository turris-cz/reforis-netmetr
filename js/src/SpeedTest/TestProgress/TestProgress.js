/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";

import useNetmetrTest from "./hooks";
import "./TestProgress.css";

TestProgress.propTypes = {
    ws: PropTypes.object.isRequired,
    asyncId: PropTypes.string,
};

export default function TestProgress({ ws, asyncId }) {
    const [data] = useNetmetrTest(ws, asyncId);

    if (!data)
        return (
            <>
                <div className="progress mb-2">
                    {/* eslint-disable-next-line jsx-a11y/control-has-associated-label */}
                    <div
                        id="progressbar"
                        className="progress-bar progress-bar-striped progress-bar-animated bg-success text-uppercase"
                        role="progressbar"
                        aria-valuenow="0"
                        aria-valuemin="0"
                        aria-valuemax="100"
                        style={{ width: "0%", fontSize: "11px" }}
                    />
                </div>
            </>
        );

    return (
        <>
            <div className="progress mb-2">
                {/* eslint-disable-next-line jsx-a11y/control-has-associated-label */}
                <div
                    id="progressbar"
                    className="progress-bar progress-bar-striped progress-bar-animated bg-success text-uppercase"
                    role="progressbar"
                    aria-valuenow={data.percent}
                    aria-valuemin="0"
                    aria-valuemax="100"
                    style={{ width: `${data.percent}%`, fontSize: "11px" }}
                >
                    {data.msg}
                </div>
            </div>
        </>
    );
}
