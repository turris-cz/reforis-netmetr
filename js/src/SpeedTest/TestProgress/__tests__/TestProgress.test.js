/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { render, act } from "foris/testUtils/customTestRender";
import { WebSockets } from "foris";

import TestProgress from "../TestProgress";

describe("<Results />", () => {
    const webSockets = new WebSockets();
    let container;
    let asFragment;
    let getByText;
    let firstRender;

    beforeEach(async () => {
        ({ container, asFragment, getByText } = render(
            <TestProgress ws={webSockets} asyncId={"test-async-id"} />
        ));
        firstRender = asFragment();
    });

    it("Should not be shown at the beginning.", () => {
        expect(firstRender).toMatchSnapshot();
    });

    it("Should render component when test is started.", () => {
        act(() =>
            webSockets.dispatch({
                module: "netmetr",
                kind: "notification",
                action: "measure_and_download_data_notification",
                data: {
                    async_id: "test-async-id",
                    percent: 5,
                    msg: "ping start",
                },
            })
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("Should be hidden when speed test is finished.", () => {
        act(() =>
            webSockets.dispatch({
                module: "netmetr",
                kind: "notification",
                action: "measure_and_download_data_notification",
                data: {
                    async_id: "test-async-id",
                    percent: 5,
                    msg: "ping start",
                },
            })
        );
        act(() =>
            webSockets.dispatch({
                module: "netmetr",
                kind: "notification",
                action: "measure_and_download_data_finished",
                data: {
                    async_id: "test-async-id",
                    passed: true,
                },
            })
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
