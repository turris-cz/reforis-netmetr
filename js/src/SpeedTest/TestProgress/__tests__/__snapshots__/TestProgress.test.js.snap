// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Results /> Should be empty when speed test is finished. 1`] = `
<DocumentFragment>
  <div
    class="progress mb-2"
  >
    <div
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="0"
      class="progress-bar progress-bar-striped progress-bar-animated bg-success text-uppercase"
      id="progressbar"
      role="progressbar"
      style="width: 0%; font-size: 11px;"
    />
  </div>
</DocumentFragment>
`;

exports[`<Results /> Should render component when test is started. 1`] = `
<DocumentFragment>
  <div
    class="progress mb-2"
  >
    <div
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="5"
      class="progress-bar progress-bar-striped progress-bar-animated bg-success text-uppercase"
      id="progressbar"
      role="progressbar"
      style="width: 5%; font-size: 11px;"
    >
      ping start
    </div>
  </div>
</DocumentFragment>
`;
